### ISSUE TITLE
Lint uses of `Vec` where arrays would do fine

### ISSUE THREAD

**Issue by clarcharr** [rust-lang-nursery/rust-clippy#2262]
For example:

```
fn something_with_slice<T>(s: &[T]) { ... }

let vals = vec![1, 2];
something_with_slice(&vals);
```

Bonus points: suggest using `ArrayVec` when the size is statically guaranteed to be less than a certain amount.

**Comment by oli-obk** [comment]
Don't we already have this? https://rust-lang-nursery.github.io/rust-clippy/master/index.html#useless_vec

Doesn't trigger in your case. So just an improvement to the existing lint required

**Comment by matthiaskrgr** [comment]
It should also catch cases like `vec![.....].iter()` where we could use a simple array:
````rust
let x: i32 = vec![1,2,3].iter().sum();
````

**Comment by y21** [comment]
I've been experimenting with the idea of improving the already existing `useless_vec` lint a little bit and came up with this idea:
- If a `vec![_]` macro invocation is found and the parent node is a local (i.e. `let ve = vec![_]`), then for each use of this local:
- Check if:
  - the use expression adjusts to a slice (e.g. calling `ve.chunks(16)`, ve adjusts to `&[_]`)
  - the parent expression is an index expression (e.g. `ve[5]`)
  - the parent expression is a method call and the method name is one of `["len", "as_ptr", "as_mut_ptr", "as_slice", "is_empty"]`
    - `as_slice` would need some special handling, we would need to also suggest removing that call
- If every use of the local passes at least one of these checks, the `vec` can (hopefully) be safely replaced with an array or a slice

This already catches *a lot* of cases in the tests alone (61) and it's going to take a while to go through all of them.

(Also, interestingly, the case @matthiaskrgr mentioned does not get caught, but changing it very slightly to immediately borrow the vec makes clippy catch it: [playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=483c940b32d6fb635a4ca93134969ae1))
